src/daemon/distribution.c

```diff
@@ -0,0 +1,402 @@
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_custom(size_t num_boundaries,
+                                   const double *custom_buckets_boundaries) {
+  bucket_t *buckets = calloc(num_boundaries + 1, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  if (num_boundaries > 0) {
+    if (custom_buckets_boundaries[0] <= 0 ||
+        custom_buckets_boundaries[0] == INFINITY) {
+      free(buckets);
+      errno = EINVAL;
+      return NULL;
+    }
+
+    buckets[0].max_boundary = custom_buckets_boundaries[0];
+
+    for (size_t i = 1; i < num_boundaries; ++i) {
+      if (custom_buckets_boundaries[i] <= 0 ||
+          custom_buckets_boundaries[i] == INFINITY ||
+          custom_buckets_boundaries[i - 1] >= custom_buckets_boundaries[i]) {
+        free(buckets);
+        errno = EINVAL;
+        return NULL;
+      }
+
+      buckets[i].max_boundary = custom_buckets_boundaries[i];
+    }
+  }
+
+  buckets[num_boundaries].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+distribution_t *distribution_new_linear(size_t num_buckets, double size) {
+  if (num_buckets == 0 || size <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  distribution_t *d = calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_linear(num_buckets, size);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+distribution_t *distribution_new_exponential(size_t num_buckets, double base,
+                                             double factor) {
+  if (num_buckets == 0 || base <= 0 || factor <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  distribution_t *d = calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  /* as in distribution_new_linear: it would be nice to check if base
+   * and factor are greater than zero, for consideration: one of them also
+   * greater than one */
+  d->buckets = bucket_new_exponential(num_buckets, base, factor);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+distribution_t *distribution_new_custom(size_t num_boundaries,
+                                        double *custom_buckets_boundaries) {
+  distribution_t *d = calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_custom(num_boundaries, custom_buckets_boundaries);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_boundaries + 1;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+static void bucket_update(bucket_t *buckets, size_t num_buckets, double gauge) {
+  int idx = (int)num_buckets - 1;
+
+  while (idx >= 0 && buckets[idx].max_boundary > gauge) {
+    buckets[idx].counter++;
+    idx--;
+  }
```

Comment:
To your mind:
```suggestion
for (size_t i = num_buckets - 1; i >= 0 && buckets[i].max_boundary > gauge) {
   buckets[i].counter++;
}
```
As we consider that distribution is not empty there won't be any issues with num_buckets - 1

Reply:
Unfortunately, if the gauge is smaller than buckets[0].max_boundary, then proposed for loop with last infinitely. Reason for it is that i is the type of size_t which is always greater than zero, so the condition i >= 0 is always true.